asmcomp/debug/available_regs.ml

```diff
@@ -0,0 +1,387 @@
+                ~provenance
+            in
+            avail_after := RD.Set.add regd (RD.Set.filter_reg !avail_after reg)
+          end
+        done;
+        Ok !avail_after
+      | Iop (Imove | Ireload | Ispill) ->
+        (* Moves are special: they enable us to propagate names.
+           No-op moves need to be handled specially---in this case, we may
+           learn that a given hard register holds the value of multiple
+           pseudoregisters (all of which have the same value).  This makes us
+           match up properly with [Liveness]. *)
+        let move_to_same_location =
+          let move_to_same_location = ref true in
+          for i = 0 to Array.length instr.arg - 1 do
+            let arg = instr.arg.(i) in
+            let res = instr.res.(i) in
+            (* Note that the register classes must be the same, so we don't
+                need to check that. *)
+            if arg.loc <> res.loc then begin
+              move_to_same_location := false
+            end
+          done;
+          !move_to_same_location
+        in
+        let made_unavailable =
+          if move_to_same_location then
+            RD.Set.empty
+          else
+            RD.Set.made_unavailable_by_clobber avail_before
+              ~regs_clobbered:instr.res
+              ~register_class:Proc.register_class
+        in
+        let results =
+          assert (Array.length instr.arg = Array.length instr.res);
+          Array.mapi (fun index result_reg ->
+              let arg_reg = instr.arg.(index) in
+              match RD.Set.find_reg_exn avail_before arg_reg with
+              | exception Not_found ->
+                RD.create_without_debug_info ~reg:result_reg
+              | arg_reg ->
+                RD.create_copying_debug_info ~reg:result_reg
+                  ~debug_info_from:arg_reg)
+            instr.res
+        in
+        Ok (RD.Set.union (RD.Set.diff avail_before made_unavailable)
+          (RD.Set.of_array results))
+      | Iop op ->
+        if M.operation_can_raise op then begin
+          augment_availability_at_raise (Ok avail_before)
+        end;
+        (* We split the calculation of registers that become unavailable after
+           a call into two parts.  First: anything that the target marks as
+           destroyed by the operation, combined with any registers that will
+           be clobbered by the operation writing out its results. *)
+        let made_unavailable_1 =
+            let regs_clobbered =
+              Array.append (Proc.destroyed_at_oper instr.desc) instr.res
+            in
+            RD.Set.made_unavailable_by_clobber avail_before ~regs_clobbered
+              ~register_class:Proc.register_class
+        in
+        (* Second: the case of OCaml to OCaml function calls.  In this case,
+           registers always become unavailable unless:
+           (a) they are "live across" the call; and/or
+           (b) they hold immediates and are assigned to the stack. *)
```

Comment:
This comment might need to explain why this is the case: the gc can move anything else.

Reply:
Fixed